### DISCUSSION TITLE
JVM crash with "Illegal threadstate encountered: 4" because of bug in RoutingModel.finalize()

### DISCUSSION THREAD

**Discussion by WiseToad-STP** [google/or-tools#2942] [Routing (and legacy CP) questions]
**What version of OR-Tools and what language are you using?**
Version: 8.1.8487
Language: Java 11

**Which solver are you using (e.g. CP-SAT, Routing Solver, GLOP, BOP, Gurobi)**
Routing Solver

**What operating system (Linux, Windows, ...) and version?**
Ubuntu 20.04.3 LTS

Using `com.google.ortools.constraintsolver.RoutingModel` class in our Java code occasionally get our JVM to crash.
Here is the part of crash log. Below this part I added the piece of your code where crashes constantly occured.

hs_err_pid.log:
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (safepoint.cpp:919), pid=1, tid=17
#  fatal error: Illegal threadstate encountered: 4
#
# JRE version: OpenJDK Runtime Environment AdoptOpenJDK (11.0.9.1+1) (build 11.0.9.1+1)
# Java VM: OpenJDK 64-Bit Server VM AdoptOpenJDK (11.0.9.1+1, mixed mode, tiered, compressed oops, g1 gc, linux-amd64)
# Core dump will be written. Default location: /opt/app_home/core
#
# If you would like to submit a bug report, please visit:
#   https://github.com/AdoptOpenJDK/openjdk-support/issues
#

---------------  S U M M A R Y ------------

Command Line: -Xmx4096m -agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=*:8787 -Djava.awt.headless=true -XX:ErrorFile=/opt/app_home/diags/err/hs_err_pid%p.log -XX:HeapDumpPath=/opt/app_home/diags/heap -XX:+HeapDumpOnOutOfMemoryError -XX:StartFlightRecording=name=disp,disk=true,maxage=3d,maxsize=3g,settings=profile.jfc -XX:FlightRecorderOptions=repository=/opt/app_home/diags/jfr,stackdepth=1024,samplethreads=true /app.jar

Host: Intel(R) Xeon(R) Gold 6240 CPU @ 2.60GHz, 6 cores, 12G, Alpine Linux v3.12
Time: Thu Dec  2 00:33:16 2021 UTC elapsed time: 30667.508093 seconds (0d 8h 31m 7s)

---------------  T H R E A D  ---------------

Current thread (0x00007f455c240800):  JavaThread "Finalizer" daemon [_thread_in_vm, id=17, stack(0x00007f4544148000,0x00007f4544249000)]

Stack: [0x00007f4544148000,0x00007f4544249000],  sp=0x00007f4544246030,  free space=1016k
Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0xf2986e]  VMError::report_and_die(int, char const*, char const*, __va_list_tag*, Thread*, unsigned char*, void*, void*, char const*, int, unsigned long)+0x37e
V  [libjvm.so+0xf2a4ff]  VMError::report_and_die(Thread*, void*, char const*, int, char const*, char const*, __va_list_tag*)+0x2f
V  [libjvm.so+0x6bfb8a]  report_fatal(char const*, int, char const*, ...)+0x10a
V  [libjvm.so+0xd1b4da]  SafepointSynchronize::block(JavaThread*)+0x8a
V  [libjvm.so+0xeb3fdc]  JavaThread::check_safepoint_and_suspend_for_native_trans(JavaThread*)+0x2fc
V  [libjvm.so+0x930ead]  ThreadStateTransition::transition_from_native(JavaThread*, JavaThreadState) [clone .constprop.266]+0x8d
V  [libjvm.so+0x931f10]  jni_DeleteGlobalRef+0x40

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  com.google.ortools.constraintsolver.mainJNI.delete_RoutingModel(J)V+0
j  com.google.ortools.constraintsolver.RoutingModel.delete()V+25
j  com.google.ortools.constraintsolver.RoutingModel.finalize()V+1
J 51457 c1 java.lang.ref.Finalizer.runFinalizer(Ljdk/internal/misc/JavaLangAccess;)V java.base@11.0.9.1 (117 bytes) @ 0x00007f454be5a92c [0x00007f454be5a240+0x00000000000006ec]
j  java.lang.ref.Finalizer$FinalizerThread.run()V+46 java.base@11.0.9.1
v  ~StubRoutines::call_stub

...
```

RoutingModel.class:
```java
package com.google.ortools.constraintsolver;

...

public class RoutingModel {

    ...

    protected void finalize() {
        this.delete();
    }

    public synchronized void delete() {
        if (this.swigCPtr != 0L) {
            if (this.swigCMemOwn) {
                this.swigCMemOwn = false;
                mainJNI.delete_RoutingModel(this.swigCPtr); // <-- calling this method occasionally get our JVM to sudden crash
            }

            this.swigCPtr = 0L;
        }

    }

    ...
```

Apparently there is bug in native code of `mainJNI.delete_RoutingModel` method.
Please fix this bug. Thank you.

**Comment by lperron** [comment]
can you try with the latest version ?

**Comment by WiseToad-STP** [comment]
Unfortunately, at the moment we cannot try with the very last version of OR Tools because we're using [ortools-java](https://github.com/magneticflux-/ortools-java) self-extractor in our project. The latest packaged version of OR Tools there is the 8.1.8487 as of Jan 07, 2021.

**Reply by Mizux** [reply]
https://search.maven.org/artifact/com.google.ortools/ortools-java/9.1.9490/jar ?

**Reply by WiseToad-STP** [reply]
We'll try, thank you. I do not know why my colleague used that unofficial solution.

**Reply by WiseToad-STP** [reply]
Now I know why we used the unofficial package. The official (you kindly suggested above) do not handle properly situations when application built as a single jar, such as Spring's bootJar. The single whole bootJar contains all the libs the application depends on, including OR-Tools jar, which in turn contains native libs needed to be unpacked and loaded to bootstrap the OR-Tools functionality. So I think the official loader works well when OR-Tools' jar comes with other jars as files in classpath needed for the application, but gives up when application wrapped in a single bootJar, as our app. The unofficial package do it best to handle any kind of such an application packaging.
Eventually, wrote my own loader. Now we using the latest version of OR-Tools as you suggested. And further, near the end of our routine when OR-Tools' RoutingModel instance used, I added the `RoutingModel.delete()` method invoke, wrapped in try-catch block to move execution of this occasionally crashing code from finalization process, executed by the JVM in postponed manner. It won't save JVM  from crash if cause of it nested inside the native code though. Maybe it adds a little bit of certainty.
So far, sitting around and watching.
I can share my loader code. Is it worth it for your project and how to do it?
